go/types/sequence_cursor.go

```diff
@@ -32,16 +49,67 @@ func (cur *sequenceCursor) getItem(idx int) sequenceItem {
 	return cur.seq.getItem(idx)
 }
 
+// sync loads the sequence that the cursor index points to.
+// It's called whenever the cursor advances/retreats to a different chunk.
 func (cur *sequenceCursor) sync() {
 	d.PanicIfFalse(cur.parent != nil)
 	cur.seq = cur.parent.getChildSequence()
+	if cur.readAheadMap != nil {
+		// if this cursor uses read-ahead, reset the cache for the next chunk
+		cur.readAheadMap = make(map[int]chan sequence)
+		cur.readAheadNext = 0
+	}
+}
+
+var readAheadParallelism = runtime.NumCPU() * 8
+
+func minInt(x, y int) int {
+	return int(math.Min(float64(x), float64(y)))
 }
 
+func maxInt(x, y int) int {
+	return int(math.Max(float64(x), float64(y)))
+}
+
+// readAhead (called when read-ahead is enabled) primes the next entries in the
+// read-ahead cache. It ensures that go routines have been allocated for reading
+// the next n entries in the current sequence. N is either readAheadParallelism
+// or the number of entries left in the sequence if smaller.
+func (cur *sequenceCursor) readAhead() {
+	d.PanicIfTrue(cur.readAheadMap == nil)
+	// the next position to be primed
+	cur.readAheadNext = maxInt(cur.idx, cur.readAheadNext)
+	// the last position to be primed
+	last := minInt(cur.idx+readAheadParallelism, cur.seq.seqLen()) - 1
+	// prime all positions between next and last
+	for i := cur.readAheadNext; i <= last; i += 1 {
+		futureSeq := make(chan sequence)
+		cur.readAheadMap[i] = futureSeq
+		go func(seq sequence, idx int) {
+			defer close(futureSeq)
+			futureSeq <- seq.getChildSequence(idx)
+		}(cur.seq, i)
+		cur.readAheadNext += 1
+	}
+}
+
+// getChildSequence retrieves the child at the current cursor position.
 func (cur *sequenceCursor) getChildSequence() sequence {
+	if cur.readAheadMap != nil {
+		// keep priming the cache
+		cur.readAhead()
+
+		// read the child from channel if it's there
+		if raChan, ok := cur.readAheadMap[cur.idx]; ok {
+			result := <-raChan
```

Comment:
With respect to readAhead (potentially) not being initiated until the first call to getChildSequence(), an optimisation here is to only readahead from index `cur.idx + 1` instead of `cur.idx` and calling `cur.seq.getChildSequence` on this thread immediately.

Because... looking at the implementation of readAhead, it seems like if cur.idx and readAheadNext are both 0 then we immediately create channels for indices `0 ... P` where P is parallelism. Then the `ok` check here will be true, and we immediately wait on the channel.

If we instead created channels for indices `1 ... P` then we wouldn't block. Hopefully this can be done without adding any complexity?